### PR TITLE
[WIP] RP Multistage Docker Build

### PR DESCRIPTION
**Pull request by bennerv**
### Which issue this PR addresses:
[VSTS 9178325](https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/9178325)

### What this PR does / why we need it:
Build the RP using a multistage docker file to eliminate the golang version dependency from our CI VMs.

### Test plan for issue:
`make image-aro` successful

### Is there any documentation that needs to be updated for this PR?
No

### CODE REVIEW

**Comment by bennerv** [comment]
A couple things to note: 
1. This approach uses a ubi8 image to build and then run the RP where the CI VMs are based on RHEL7.  This is because the  `gpgme-devel` library isn't included in ubi7/rhel7 based images, but it is in ubi8/rhel8 images.
2. Other OpenShift-based approaches use the CI images (`registry.ci.openshift.org/openshift/release:golang-1.15`) to build their applications which appears to be based on CentOS.  I ran into library issues building the RP on Fedora and attempting to run it on ubi8, so I tried to keep the builder and runner containers as close to each other as possible.

**Comment by bennerv** [comment]
So the version of docker installed in our CI VMs doesn't support multistage builds (too early of a version).  I'll go through the work of updating `pkg/deploy/generator/resources.go` to try to run a newer version of docker by using the CentOS packages [here](https://docs.docker.com/engine/install/centos/) as the RHEL7 package is old and not updated.

**Comment by jim-minter** [comment]
Thanks for this @bennerv .

Although I'd love to go to RHEL 8 / ubi8, I prefer being standardised, and we need 7 because azsecpack still doesn't support 8.  I'm -1 on our current use of ubi8 in this repo and would rather it was 7.

gpgme-devel is available in RHEL7 - I guess in Optional - we install it on RHEL7 in CI here: https://github.com/Azure/ARO-RP/blob/master/pkg/deploy/generator/resources.go#L1956.  Probably the difficulty is that you need the build host to be subscribed, which is doable but fiddly on Fedora, and perhaps horrid on Macs?

I wonder whether we end up needing to push a builder container for the benefit of Mac-using-engineers?

I'm -100 on introducing newer docker into our build chain - it adds a dependency on another corporate entity.  I would be OK with using RHEL7 podman across the board if it works though.

TL;DR: I think this PR needs to be closed for now for a few smaller(?) prerequisites:
1) ubi8 -> rhel7
2) docker -> podman in RP VM + CI VM + Makefile in a way which is OK for Mac folks
3) try again

**Comment by bennerv** [comment]
Closing in place of changes to make outlined above

**Comment by bennerv** [comment]
@jim-minter I think this is in a better state filling a few of the comments you mentioned.  

A few questions and comments to note:
1. I went with a builder image approach I figured we would store in the Azure container registry.  
2. The repositories don't exist for the builder image.  Not sure if I have permission to do that.  
3. I would like to move the image build out of the normal RP image build and leave it tagged with the golang version that is used by the builder image.  So it should be manually triggered as a pipeline instead of built and pushed every time by the pipeline.
4. I'm not sure if `podman` is installed by default in rhel7.  We might need to have it installed and redeploy the CI VMSS in the future if we want to go with this.
5. I removed some image pulls being done in the makefile commands.  More than happy to add those back in, was just wondering why they were there before the build in the first place? (Maybe fail early instead of attempt to push an image it can't)

Leaving it in draft until we address one or two of the above and additional comments.
